core/src/avm2/array.rs

```diff
@@ -85,44 +291,171 @@ impl<'gc> ArrayStorage<'gc> {
+        match self {
+            ArrayStorage::Dense { .. } => {}
+            ArrayStorage::Sparse { storage, .. } => {
+                if storage.is_empty() {
+                    *self = ArrayStorage::Dense {
+                        storage: Vec::new(),
+                        occupied_count: 0,
+                    };
+                }
+            }
+        }
     }
 
     /// Push an array hole onto the end of this array.
     pub fn push_hole(&mut self) {
-        self.storage.push(None)
+        match self {
+            ArrayStorage::Dense { storage, .. } => {
+                storage.push(None);
+                self.maybe_convert_to_sparse();
+            }
+            ArrayStorage::Sparse { length, .. } => {
+                *length += 1;
+            }
+        }
     }
 
     /// Pop a value from the back of the array.
     ///
     /// This method preferrentially pops non-holes from the array first. If a
     /// hole is popped, it will become `undefined`.
     pub fn pop(&mut self) -> Value<'gc> {
-        let mut non_hole = None;
+        match self {
+            ArrayStorage::Dense {
+                storage,
+                occupied_count,
+            } => {
+                let non_hole = storage
+                    .iter()
+                    .enumerate()
+                    .rposition(|(_, item)| item.is_some());
 
-        for (i, item) in self.storage.iter().enumerate().rev() {
-            if item.is_some() {
-                non_hole = Some(i);
-                break;
+                if let Some(non_hole) = non_hole {
+                    *occupied_count -= 1;
+                    let value = storage.remove(non_hole).unwrap();
+                    self.maybe_convert_to_sparse();
+                    value
+                } else {
+                    storage.pop().unwrap_or(None).unwrap_or(Value::Undefined)
+                }
             }
-        }
+            ArrayStorage::Sparse { storage, length } => {
+                if *length == 0 {
+                    return Value::Undefined;
+                }
 
-        if let Some(non_hole) = non_hole {
-            self.storage.remove(non_hole).unwrap()
-        } else {
-            self.storage
-                .pop()
-                .unwrap_or(None)
-                .unwrap_or(Value::Undefined)
+                let mut non_hole = None;
+
+                let storage_clone = storage.clone();
+                if let Some((i, _)) = storage_clone.iter().next_back() {
+                    non_hole = Some(i);
+                }
+                if let Some(non_hole) = non_hole {
+                    *length -= 1;
+                    let value = storage.remove(non_hole).unwrap();
+                    self.maybe_convert_to_dense();
+                    value
+                } else {
+                    let value = storage
+                        .get(&(*length - 1))
+                        .copied()
+                        .unwrap_or(Value::Undefined);
```

Comment:
do I understand correctly that this only runs if the array is made entirely of holes? In that case, I think both this and the code for Dense (as in, the pre-existing one) sound overly complex, as `storage.get` must always be `None`; unless I'm missing some logic here.